integrate/simpsons_test.go

```diff
@@ -0,0 +1,130 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+	"testing"
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func TestSimpsons(t *testing.T) {
+	const N = 1e6
+	x := floats.Span(make([]float64, N), 0, 1)
+	for i, test := range []struct {
+		x    []float64
+		f    func(x float64) float64
+		want float64
+	}{
+		{
+			x: []float64{0.0, 0.1, 1.0},
+			f: func(x float64) float64 {
+				switch {
+				case x >= 1e-31 || x <= 1.0:
+					return 1.0
+				default:
+					return 0
+				}
+			},
+			want: 1,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x },
+			want: 0.5,
+		},
+		{
+			x:    floats.Span(make([]float64, N), -1, 1),
+			f:    func(x float64) float64 { return x },
+			want: 0,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x + 10 },
+			want: 10.5,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return 3*x*x + 10 },
+			want: 11,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return math.Exp(x) },
+			want: 1.7182818284591876,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 2,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, 0.5*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 1,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 0,
+		},
+	} {
+		y := make([]float64, len(test.x))
+		for i, v := range test.x {
+			y[i] = test.f(v)
+		}
+		v := Simpsons(test.x, y)
+		if !floats.EqualWithinAbs(v, test.want, 1e-12) {
+			t.Errorf("test #%d: got=%v want=%f\n", i, v, float64(test.want))
+		}
+	}
+}
+
+func TestSimpsonsPanics(t *testing.T) {
```

Comment:
I don't think we need to test this. We don't do such tests anywhere else in Gonum.

Reply:
I think tests for panics are really important. If I mix-up upper and lower bounds for example, I feel like those sorts of things should be introduced.

Reply:
I think the cost of this outweighs any potential benefits. It clutters the CI output and is more code to maintain and keep in sync with the implementation for practically no benefit. @kortschak may have more argument for why we don't need this.

Reply:
The only time we check for panics is when we are testing non-obvious logic. This happens in mat, graph and lapack. We don't check for this kind of thing in stat in other place nor in floats.

In general adding tests is a way to write in a potentially clearer way an assertion about behaviour. The `validateInterval` helper is already simple enough (possibly simpler if the switch were placed in the function proper) that it is obviously correct.